src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1573,6 +1576,430 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
+            txBundle.getTx()->sorobanResources().footprint.readWrite;
+        for (auto const& lk : readWrite)
+        {
+            if (!isSorobanEntry(lk))
+            {
+                continue;
+            }
+
+            auto const& ttlKey = getTTLKey(lk);
+            auto extIt = readOnlyTtlExtensions.find(ttlKey);
+            if (extIt != readOnlyTtlExtensions.end() && extIt->second != 0)
+            {
+                // "Commit" all RO cumulative bumps now that the key is in a
+                // readWrite set
+                auto it = entryMap.find(ttlKey);
+                if (it != entryMap.end())
+                {
+                    releaseAssertOrThrow(it->second.mLedgerEntry);
+
+                    it->second.mLedgerEntry->data.ttl().liveUntilLedgerSeq =
+                        std::min(
+                            sorobanConfig.stateArchivalSettings().maxEntryTTL,
+                            add_sat(it->second.mLedgerEntry->data.ttl()
+                                        .liveUntilLedgerSeq,
+                                    extIt->second));
+
+                    // Mark as dirty so this entry gets written.
+                    it->second.isDirty = true;
+                }
+                extIt->second = 0;
+            }
+        }
+
+        auto res = txBundle.getTx()->parallelApply(
+            entryMap, config, sorobanConfig, ledgerInfo,
+            txBundle.getResPayload(), sorobanMetrics, txSubSeed,
+            txBundle.getEffects());
+
+        if (res.mSuccess)
+        {
+            UnorderedMap<LedgerKey, bool> isReadOnlyTTLMap;
+            for (auto const& ro :
+                 txBundle.getTx()->sorobanResources().footprint.readOnly)
+            {
+                if (!isSorobanEntry(ro))
+                {
+                    continue;
+                }
+                isReadOnlyTTLMap.emplace(getTTLKey(ro), true);
+            }
+            for (auto const& rw :
+                 txBundle.getTx()->sorobanResources().footprint.readWrite)
+            {
+                if (!isSorobanEntry(rw))
+                {
+                    continue;
+                }
+                isReadOnlyTTLMap.emplace(getTTLKey(rw), false);
+            }
+
+            // now apply the entry changes to entryMap
+            for (auto const& entry : res.mModifiedEntryMap)
+            {
+                auto const& lk = entry.first;
+                auto const& updatedLe = entry.second;
+
+                auto it = entryMap.find(lk);
+                releaseAssertOrThrow(it != entryMap.end());
+
+                auto opType =
+                    txBundle.getTx()->getRawOperations().at(0).body.type();
+
+                auto isReadOnlyTTLIt = isReadOnlyTTLMap.find(lk);
+
+                if (opType != RESTORE_FOOTPRINT && lk.type() == TTL &&
+                    it->second.mLedgerEntry && updatedLe &&
+                    isReadOnlyTTLIt != isReadOnlyTTLMap.end() &&
```

Comment:
The entry can't be missing from the map, can it? Maybe an assertion would be more appropriate to ensure we didn't mess up anywhere.

Reply:
Added an assert. I still need to go over all of the asserts to make sure they're safe and the ones that throw aren't caught when they shouldn't be.